### PR TITLE
NIFI-12709 Added ability to get more attributes for zip files as well as created new attributes to get for both tar and zip files.

### PR DESCRIPTION
**Pull request by dan-s1**
# Summary
Three attributes `file.lastMetadataChange`, `file.lastAccessTime` and `file.size` have been added and other values for existing attributes have been made available (`file.creationTime` and `file.permissions`) . Below is a chart of all attributes the UnpackContent processor now has and the ones in bold are the ones which have been added. For tar files the `file.creationTime` before this PR used the `file.lastModifiedTime` but now if the  actual `file.creationTime` is available it is used and only defaults to using the `file.lastModifiedTime` if the `file.creationTime` is not available (this same logic is now employed for unencrypted zip files). For enrypted zip files the value for `file.lastModifiedTime` is used for the attribute `file.lastModifiedTime` and `file.creationTime`.

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<html>
<body>

Java Variable | Attribute Name | Tar | Unencrypted Zip | Encrypted Zip
-- | -- | -- | -- | --
FILE_LAST_MODIFIED_TIME_ATTRIBUTE | file.lastModifiedTime | Y | Y | Y
FILE_CREATION_TIME_ATTRIBUTE | file.creationTime | **Y** | **Y** | **Y**
**FILE_LAST_METADATA_CHANGE_ATTRIBUTE** | **file.lastMetadataChange** | **Y** | N | N
**FILE_LAST_ACCESS_TIME_ATTRIBUTE** | **file.lastAccessTime** | **Y** | **Y** | N
FILE_OWNER_ATTRIBUTE | file.owner | Y | N | N
FILE_GROUP_ATTRIBUTE | file.group | Y | N | N
**FILE_SIZE_ATTRIBUTE** | **file.size** | **Y** | **Y** | **Y**
FILE_PERMISSIONS_ATTRIBUTE | file.permissions | Y | **Y** | N
FILE_ENCRYPTION_METHOD_ATTRIBUTE | file.encryptionMethod | N | Y | Y

</body>

</html>

Currently Zip4j the implementation used for encrypted zip files does not have access methods for `file.permissions` `file.creationTime` and `file.lastAccessTime`. A [request](https://github.com/srikanth-lingala/zip4j/issues/557) has been made to add these.
 
[NIFI-12709](https://issues.apache.org/jira/browse/NIFI-12709)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/UnpackContent.java [hunk 1]
```diff
@@ -470,22 +500,47 @@ protected boolean isFileEntryMatched(final boolean directory, final String fileN
...
-                                FRAGMENT_INDEX, String.valueOf(++fragmentIndex)
-                        ));
+                        final Map<String, String> attributes = new HashMap<>();
+                        attributes.put(CoreAttributes.FILENAME.key(), file.getName());
+                        attributes.put(CoreAttributes.PATH.key(), parentDirectory);
+                        attributes.put(CoreAttributes.MIME_TYPE.key(), OCTET_STREAM);
+                        attributes.put(FILE_ENCRYPTION_METHOD_ATTRIBUTE, metadata.encryptionMethod().toString());
+                        attributes.put(FILE_SIZE_ATTRIBUTE, String.valueOf(metadata.uncompressedSize()));
+
+                        String timeAsString = null;
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/UnpackContent.java:517]: Reusing this variable makes the code difficult to follow. I recommend avoiding reuse and defining specific variables inside the conditional blocks as needed.

**Review by exceptionfactory — COMMENTED** [review]
Thanks for the improvements @dan-s1. The changes look straightforward in general, I just noted one recommendation to make the code a bit clearer.

One additional note, it looks like it should be possible to update the unit test to check for additional attributes, such as `file.size`, and some of the TAR attributes.

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/UnpackContent.java [hunk 2]
```diff
@@ -437,6 +463,11 @@ public void unpack(final ProcessSession session, final FlowFile source, final Li
             }
         }
 
+        private record ZipInputStreamMetadata(boolean directory, String zipEntryName, EncryptionMethod encryptionMethod,
+                                               Instant creationTime, Instant lastModifiedDate, Instant lastAccessDate, int mode,
+                                               long uncompressedSize) {
+        }
```
Comment (exceptionfactory) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/UnpackContent.java:469]: Recommend renaming this `record` to `ZipEntryMetadata` as it only applies to a single entry. Also recommend listing each property on its own line for clarity.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the updates @dan-s1. This looks closer to completion. In the course of closer review, it seems worth reconsidering the introduction of the new `record` for metadata. Instead of creating a new object for every entry, what do you think about creating separate methods for creating the new `Map` of attributes? Common properties could be passed as arguments, and then the standard `ZipEntry` could set additional values, like file mode and access time, avoiding the intermediate object creation for every entry.

**Review by exceptionfactory — APPROVED** [review]
Thanks for making the adjustments @dan-s1, the latest version looks good. +1 merging

**Comment by dan-s1** [comment]
@exceptionfactory Thanks for accepting this PR and merging. I just wanted to note I was careful to make sure to instantiate one java.util.Map for creating attributes for both tar and zip entries to be mindful of the issue brought up in [NIFI-13288](https://issues.apache.org/jira/browse/NIFI-13288)

**Comment by exceptionfactory** [comment]
> @exceptionfactory Thanks for accepting this PR and merging. I just wanted to note I was careful to make sure to instantiate one java.util.Map for creating attributes for both tar and zip entries to be mindful of the issue brought up in [NIFI-13288](https://issues.apache.org/jira/browse/NIFI-13288)

Thanks, I noted that when reviewing the implementation, that's helpful.

**Pull request closed without merging**